File: 15_3Sum.py
class Solution(object):
    def threeSum(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        result = []
        nums.sort()
        for first_index in xrange(len(nums)):
            if first_index - 1 >= 0 and nums[first_index] == nums[first_index-1]:
                continue
            start, end = first_index + 1, len(nums)-1
            while start < end :
                tmp_sum = nums[first_index] + nums[start] + nums[end] 
                
                if tmp_sum == 0:
                    result.append([nums[first_index], nums[start], nums[end]])
                    start += 1
                    end -= 1
                    while start < end and nums[start] == nums[start-1]:
                        start+=1
                    
                elif tmp_sum < 0:
                    start += 1
                else:
                    end -= 1
                    
        return result

File: solutions/15_3Sum.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var threeSum = function(nums) {
    let result = [];
    let sum = 0;
	let sorted_arr = nums.sort((a,b)=>{
		return a-b;
	})
	for(let i = 0; i<sorted_arr.length;i++){
        if(i-1>=0 && nums[i] == nums[i-1]){continue}
		let j = i+1;
		let k = sorted_arr.length-1;
		while(j<k){
			if(sorted_arr[i]+sorted_arr[j]+sorted_arr[k]===sum){
				result.push([sorted_arr[i],sorted_arr[j],sorted_arr[k]]);
				j++;
				while(j<k && sorted_arr[j]===sorted_arr[j-1]) { j++ }
			} else if(sorted_arr[i]+sorted_arr[j]+sorted_arr[k] < sum){
				j++;
			} else {
				k--;
			} 
		}
	}
	return result
    
};